docs/Account/methods.md

```diff
@@ -0,0 +1,284 @@
+const nonce = await smartAccount.getNonce();
+console.log(nonce.toNumber());
+```
+**Parameters**
+- nonceKey(`number`): one can also pass the optional parameter nonceKey in the case of two-dimensional nonces
+
+**Returns**
+
+- address (`Promise<BigNumber>`): A Promise resolving to BigNumber indicating the nonce of the smart account.
+
+### index
+
+This is used to retrieve the index of the current active smart account.
+
+**Usage**
+```jsx
+const index = smartAccount.index;
+```
+
+**Returns**
+- index (`number`): A number indicating the index of current active smart account.
+
+## UserOp Methods
+
+### buildUserOp( )
+This method is used for configuring and setting up all properties of the `userOp` object.  It involves preparing and populating the userOp with necessary information and parameters required for execution by bundler.
+
+**Usage**
+
+For example, in the context of creating a userOp for an `addComment` transaction, an instance of the contract is created, then a basic transaction object is created that holds the necessary address and data from the transaction. Finally, a userOp is created using the Smart Account's `buildUserOp` method⁠.
+
+```jsx
+const contractAddress = "contract address";
+const provider = new ethers.providers.JsonRpcProvider("rpc url");
+
+const blogContract = new ethers.Contract(
+  contractAddress,
+  abi, // contract abi
+  provider,
+);
+
+const createComment =
+  await blogContract.populateTransaction.addComment("comment");
+
+const tx1 = {
+  to: contractAddress,
+  data: createComment.data,
+};
+
+const userOp = await smartAccount.buildUserOp([tx1]);
+```
+
+**Parameters**
+- transactions (`Transaction[]`, required): The required argument  is an array of transactions. You can pass multiple transactions into a userOp if you would like to batch them together into one transaction.
+- buildUseropDto (`BuildUserOpOptions`): One can also pass these options to customize how a userOp is built.
+
+  ```ts
+  type BuildUserOpOptions = {
+    overrides?: Overrides;
+    skipBundlerGasEstimation?: boolean;
+    params?: ModuleInfo;
+    nonceOptions?: NonceOptions; 
+    forceEncodeForBatch?: boolean;
+    paymasterServiceData?: SponsorUserOperationDto;
+  }
+  ```
+  Let's look at each of these params:
+  1. overrides (`Overrides`): one can override any of the values of the userOp when it is being constructed.
+
+      ```ts
+      type BigNumberish = BigNumber | Bytes | bigint | string | number;
+
+      type Overrides = {
+        callGasLimit?: BigNumberish;
+        verificationGasLimit?: BigNumberish;
+        preVerificationGas?: BigNumberish;
+        maxFeePerGas?: BigNumberish;
+        maxPriorityFeePerGas?: BigNumberish;
+        paymasterData?: string;
+        signature?: string; // todo - add the definition
+      }
+      ```
+
+  2. skipBundlerGasEstimation (`boolean`): This boolean param can be set to false to skip some Gas Estimations on the bundler which can help with reducing latency on requests.
```

Comment:
need more information on this. @AmanRaj1608 can you help draft this please

Reply:
**skipBundlerGasEstimation (boolean)**: This parameter allows you to manage gas estimations more efficiently depending on the nature of your transactions. By default, it's set to true, which means if a paymaster is present, gas estimations are done on the paymaster side to facilitate gasless transactions. Here's how you should configure it based on your use case:

1. Gasless Paymaster Flow (All Methods Gasless on Dashboard): If you are utilizing a gasless transaction with all methods set as gasless on the dashboard, you should pass the skipBundlerGasEstimation as true. This is typically the default behavior, emphasizing that the gas estimations are handled by the paymaster. Here's how you can configure it:

```js
let partialUserOp = await biconomySmartAccount.buildUserOp([transaction], {
  paymasterServiceData: {
    mode: PaymasterMode.SPONSORED,
  },
  // No need to explicitly set skipBundlerGasEstimation: true, as it's true by default for paymaster flow
});
```
2. Mixed or Non-Gasless Transactions: In scenarios where you have mixed transactions (some are gasless and some are not) or all transactions are not gasless, it's more efficient to estimate gas using the bundler first. After building the user operation, you should then call getPaymasterAndData to finalize the gas parameters. Set skipBundlerGasEstimation as false to ensure the bundler handles the initial gas estimations:

```js
// When not all methods are gasless or you want to use bundler estimations
let userOp = await biconomySmartAccount.buildUserOp([transaction], {
  skipBundlerGasEstimation: false, // Explicitly set to false to use bundler for initial gas estimations
});
// ...After building user operation, proceed to get Paymaster and Data as needed
```

By configuring skipBundlerGasEstimation appropriately, you ensure that your transactions are processed efficiently and in accordance with the intended gas handling approach, whether through a gasless paymaster flow or via bundler estimations.

Reply:
what does user optimise on by getting the gas calculations done by paymaster vs bundler?

Reply:
so generally you build partial user op , use the bundler to do gas estimations
paymasterAndData (we ll call it pnd from now on) is not populated yet
then generally you can paymaster to get paymasterAndData.

at this time paymaster services could / should do gas estimations again cause using paymaster impacts gas values

so the idea is if you are sure you’re going to use sponsored mode with paymaster, and since paymaster service can do estimations… in this case instead of bundler make paymaster call with calculateGasLimits: true (default behavior) and get gas estimations + pnd also

improves on multiple calls and latency in this particular case